pkg/p2p/wire/message/reduction.go

```diff
@@ -0,0 +1,209 @@
+package message
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/header"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/encoding"
+	"github.com/dusk-network/dusk-blockchain/pkg/util"
+	"github.com/dusk-network/dusk-crypto/bls"
+	"github.com/dusk-network/dusk-wallet/v2/key"
+)
+
+type (
+	// Reduction is one of the messages used in the consensus algorithms. As
+	// such it encapsulates a header.Header to allow the Coordinator to
+	// correctly enforce the sequence of state changes expressed through the algorithm
+	Reduction struct {
+		hdr        header.Header
+		SignedHash []byte
+	}
+)
+
+// New returns and empty Reduction event.
+func NewReduction(hdr header.Header) *Reduction {
+	return &Reduction{
+		hdr:        hdr,
+		SignedHash: make([]byte, 33),
+	}
+}
+
+func (r Reduction) String() string {
+	var sb strings.Builder
+	sb.WriteString(r.hdr.String())
+	sb.WriteString(" signature' ")
+	sb.WriteString(util.StringifyBytes(r.SignedHash))
+	sb.WriteString("'")
+	return sb.String()
+}
+
+// Header is used to comply to consensus.Message
+func (r Reduction) State() header.Header {
+	return r.hdr
+}
+
+// Header is used to comply to consensus.Message
+func (r Reduction) Sender() []byte {
+	return r.hdr.Sender()
+}
+
+// Equal is used to comply to consensus.Message
+func (r Reduction) Equal(msg Message) bool {
+	m, ok := msg.Payload().(Reduction)
+	return ok && r.hdr.Equal(m.hdr) && bytes.Equal(r.SignedHash, m.SignedHash)
+}
+
+func UnmarshalReductionMessage(r *bytes.Buffer, m SerializableMessage) error {
+	bev := NewReduction(header.Header{})
+	if err := header.Unmarshal(r, &bev.hdr); err != nil {
+		return err
+	}
+
+	if err := UnmarshalReduction(r, bev); err != nil {
+		return err
+	}
+
+	m.SetPayload(bev)
+	return nil
+}
+
+// Unmarshal unmarshals the buffer into a Reduction event.
+func UnmarshalReduction(r *bytes.Buffer, bev *Reduction) error {
+	if err := header.Unmarshal(r, &bev.hdr); err != nil {
```

Comment:
When called from `UnmarshalReductionMessage`, the header is unmarshalled twice, leading to unnecessary EOF errors.